drools-compiler/src/main/java/org/drools/compiler/kie/builder/impl/KieRepositoryImpl.java

```diff
@@ -189,7 +189,11 @@ private KieModule checkClasspathForKieModule(ReleaseId releaseId) {
     }
 
     private KieModule loadKieModuleFromMavenRepo(ReleaseId releaseId, PomModel pomModel) {
-        return KieScannerHolder.kieScanner.loadArtifact( releaseId, pomModel );
+        KieModule kieModule;
+        synchronized (kieModuleRepo) { // Make sure kieModuleRepo lock is acquired before kieScanner lock
+            kieModule = KieScannerHolder.kieScanner.loadArtifact(releaseId, pomModel);
+        }
```

Comment:
Ensuring the lock order is the basic solution for a deadlock. As I see, problematic scenario is only this place, so I solved with this small fix.

An alternative approach would be to create a `ReentrantLock` and lock it for both `kieModuleRepo` and `KieScannerHolder.kieScanner` all usage. But it would require more code changes. Which do you prefer? > @mariofusco

Reply:
We should replace all these `synchronized` block with `ReentrantLock`s, but this change can be done only on the main branch and it is not urgent however.

Reply:
Thanks, I'll do that for the main branch.